### PR TITLE
Expand Bindings

### PR DESCRIPTION
**Pull request by kaendfinger**
I am currently working on binding almost everything I can. I will try to summarize things as I go.

**NOTE**: Highly recommend you squash and merge this when it is ready.

## Changes

- `Odb::new`: Constructor for creating an empty ODB.
- `Repository::from_odb`: Constructor for creating a repository from an ODB.
- `Odb::zero`: Adds a constructor for a zero'd Oid.
- `Oid::hash_file`: Binds the [git_odb_hashfile](https://libgit2.github.com/libgit2/#v0.26.0/group/odb/git_odb_hashfile) function.
- `Oid::hash_object`: Binds the [git_odb_hash](https://libgit2.github.com/libgit2/#v0.26.0/group/odb/git_odb_hash) function.
- Creates a representation of the `git_odb_backend` and `git_refdb_backend` structures in the raw crate.
- Adds the `Odb.refresh` method binding.
- Adds the `Odb.exists_prefix` method binding.
- Adds the `Odb.write` method binding.
- Adds the `Odb.read_header` method binding.
- Binds `Repository.is_worktree()`
- Creates a constructor `Repository::open_bare` to open a bare repository.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the progress here! Lemme know when this is ready and I'll take a look

**Comment by kaendfinger** [comment]
@alexcrichton I believe it is safe to look at this now. I plan on adding more in the coming weeks.

File: src/odb.rs [hunk 1]
```diff
@@ -78,10 +89,107 @@ impl<'repo> Odb<'repo> {
...
+        unsafe {
+            try_call!(raw::git_odb_add_alternate(self.raw, backend.raw(), priority));
+            Ok(())
+        }
+    }
+
+    /// Adds an alternate disk backend to the object database.
+    pub fn add_disk_alternate(&self, path: &str) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_odb_add_disk_alternate(self.raw, path.as_ptr() as *mut _));
```
Comment (alexcrichton) [inline src/odb.rs]: This probably wants to use a CString of some kind, right?

File: src/odb_backend.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+impl OdbBackendHolder {
+    /// Creates an object database backend for a loose object directory.
+    pub fn loose(objects_dir: &str, compression_level: i32,
+                 do_fsync: bool, dir_mode: u32, file_mode: u32) -> Result<OdbBackendHolder, Error> {
+        ::init();
+        unsafe {
+            let mut out = ptr::null_mut();
+            try_call!(raw::git_odb_backend_loose(
+                &mut out as *mut _,
+                objects_dir.as_ptr() as *const libc::c_char,
```
Comment (alexcrichton) [inline src/odb_backend.rs]: CString here as well?

File: src/odb_backend.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /// Creates an object database backend that uses pack in a given directory.
+    pub fn pack(objects_dir: &str) -> Result<OdbBackendHolder, Error> {
+        ::init();
+        unsafe {
+            let mut out = ptr::null_mut();
+            try_call!(raw::git_odb_backend_pack(
+                &mut out as *mut _,
+                objects_dir.as_ptr() as *const libc::c_char
```
Comment (alexcrichton) [inline src/odb_backend.rs]: CString?

File: src/odb_backend.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /// Creates an object database backend from a single pack.
+    pub fn one_pack(index_file: &str) -> Result<OdbBackendHolder, Error> {
+        ::init();
+        unsafe {
+            let mut out = ptr::null_mut();
+            try_call!(raw::git_odb_backend_one_pack(
+                &mut out as *mut _,
+                index_file.as_ptr() as *const libc::c_char
```
Comment (alexcrichton) [inline src/odb_backend.rs]: CString?

File: src/odb_backend.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
+use {raw, Buf, Error, Repository};
+
+use std::ptr;
+
+use util::Binding;
+use libc;
+
+/// A structure to represent a git object database backend.
+pub struct OdbBackendHolder {
```
Comment (alexcrichton) [inline src/odb_backend.rs]: Hm I wonder if we perhaps want to make this a trait or a trait object at some point? It looks like this is similar to the `Transport` trait perhaps in git2-rs?

File: src/oid.rs [hunk 6]
```diff
@@ -44,6 +46,50 @@ impl Oid {
         }
     }
 
+    /// Creates an all zero Oid structure.
+    pub fn zero() -> Oid {
+        ::init();
```
Comment (alexcrichton) [inline src/oid.rs]: It's probably ok to skip the `::init` here as libgit2 isn't called.

File: src/repo.rs [hunk 7]
```diff
@@ -368,6 +396,47 @@ impl Repository {
...
+    /// Set the active namespace for this repository.
+    pub fn set_namespace(&self, namespace: &str) -> Result<(), Error> {
+        self.set_namespace_bytes(namespace.as_bytes())
+    }
+
+    /// Set the active namespace for this repository as a byte array.
+    pub fn set_namespace_bytes(&self, namespace: &[u8]) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_repository_set_namespace(self.raw,
+                                                        namespace.as_ptr() as *mut c_char));
```
Comment (alexcrichton) [inline src/repo.rs]: CString?

**Comment by kaendfinger** [comment]
I'll try to address this ASAP

File: src/odb_backend.rs (hunk 5 above)
Comment (kaendfinger) [reply src/odb_backend.rs]: Yes, it is very similar. I attempted a trait implementation, however it quickly became a mess. Is there a way to simultaneously have the structure (backed by the public functions), and the trait to implement custom backends?

Comment (alexcrichton) [reply src/odb_backend.rs]: Yeah doing it as a trait I think will definitely be a relatively significant investment in the sense that it'll take some design. Perhaps that part could be left out from this PR for now?

Comment (kaendfinger) [reply src/odb_backend.rs]: @alexcrichton I'd definitely like to hold off on the trait, however I'd prefer to be able to keep the OdbBackendHolder (or at least a way to use the functionality it provides) as I am using it in a project, but that's entirely up to you.

Comment (alexcrichton) [reply src/odb_backend.rs]: Oh my worry is that the current implementation is not memory safe, so we could just address those sorts of questions in a future PR

**Comment by kaendfinger** [comment]
That's fine then, I can go without it :)

**Comment by kaendfinger** [comment]
@alexcrichton I have removed all references to the backend bindings, and am currently working on a trait based implementation, but that will be for a later PR.

**Comment by kaendfinger** [comment]
@alexcrichton Also a reminder to squash the changes before merging, I have quite a few commits :)

**Comment by alexcrichton** [comment]
Thanks @kaendfinger!

**Pull request merged**